plugins/uefi/fu-uefi-device.c

```diff
@@ -393,6 +395,78 @@ fu_uefi_device_write_update_info (FuUefiDevice *self,
 	return TRUE;
 }
 
+static gboolean
+fu_uefi_device_is_esp_mounted (FuDevice *device, GError **error)
+{
+	const gchar *esp_path = fu_device_get_metadata (device, "EspPath");
+	g_autofree gchar *contents = NULL;
+	g_auto(GStrv) lines = NULL;
+	gsize length;
+
+	if (esp_path == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "EFI System partition is not defined");
+		return FALSE;
+	}
+
+	if (!g_file_get_contents ("/proc/mounts", &contents, &length, error))
+		return FALSE;
+	lines = g_strsplit (contents, "\n", 0);
+
+	for (guint i = 0; lines[i] != NULL; i++) {
+		if (lines[i] != NULL && g_strrstr (lines[i], esp_path))
+			return TRUE;
+	}
+	g_set_error (error,
+		     FWUPD_ERROR,
+		     FWUPD_ERROR_NOT_SUPPORTED,
+		     "EFI System partition %s is not mounted",
+		     esp_path);
+	return FALSE;
+}
+
+static gboolean
+fu_uefi_device_prepare (FuDevice *device,
+			FwupdInstallFlags flags,
+			GError **error)
+{
+	const gchar *esp_path = fu_device_get_metadata (device, "EspPath");
+
+	/* mount the partition somewhere */
+	if (fu_uefi_udisks_objpath (esp_path)) {
+		FuUefiDevice *self = FU_UEFI_DEVICE (device);
+		g_autofree gchar *path = NULL;
+		path = fu_uefi_udisks_objpath_mount (esp_path, error);
+		if (path == NULL)
+			return FALSE;
+		g_debug ("Mounted ESP at %s", path);
+		self->esp_obj_file = g_strdup (esp_path);
+		fu_device_set_metadata (device, "EspPath", path);
```

Comment:
I'm getting a bit confused with all the various places the ESP path is stored. I think this makes it three! Could we remove the one in the plugin and just use "EspPath` and then just set something like `self->auto_mounted = TRUE` so that it's only stored in the `FuUefiDevice` object and only set in `_prepare()`?

Reply:
That was one of my ideas along the way but it turns out that they all have purpose.

The plugin one is needed so that proxy created devices get an esp set as well. I guess the other alternative is to re detect any time a new device is made.

Reply:
Oh but moving it to prepare, I see what you mean. Yeah I think that can work.